### PR TITLE
feat: offer push notifications for slow responses

### PR DESCRIPTION
**Pull request by sachaservan**
## Summary

When an assistant response has been pending for more than 5 seconds, a banner appears above the chat input offering to notify the user when the response is ready. Clicking **Notify** requests notification permission, registers the browser with FCM, and asks the controlplane to push when the stream completes — so the user can switch tabs or close the window.

Companion controlplane PR: https://github.com/tinfoilsh/controlplane/pull/545 — adds the storage, watch endpoints, and FCM fan-out.

## How it works

- **Gating falls out of chat recovery**: the banner only appears when the stream has an active EHBP recovery session, which already requires a signed-in user, non-temporary chat, and history storage on. Anonymous users never see it. Also requires push support in the browser and permission not denied.
- `use-chat-messaging` publishes the live recovery session ID per chat (`active-stream-sessions.ts`), handling retry session swaps.
- `useStreamNotifyBanner` starts a 5s timer (`NOTIFY_BANNER_DELAY_MS`) once a session is pending; dismissal is per-session so a retried stream can offer again; confirmation auto-hides.
- `public/firebase-messaging-sw.js` is a plain Web Push service worker (data-only messages, no SDK import) registered under a `/push/` scope so it never intercepts page fetches. It:
  - suppresses the OS notification when the user is already viewing that chat (controlplane always sends; the worker decides)
  - focuses or opens `/chat/{chatId}` on click (local chats deep-link to `/chat/local/{chatId}`)
- CSP allows `fcmregistrations.googleapis.com` / `firebaseinstallations.googleapis.com`.

## Configuration

Push stays disabled (banner never shows) until the Firebase web config is set:
`NEXT_PUBLIC_FIREBASE_API_KEY`, `..._AUTH_DOMAIN`, `..._PROJECT_ID`, `..._MESSAGING_SENDER_ID`, `..._APP_ID`, and `NEXT_PUBLIC_FIREBASE_VAPID_KEY` (Web Push certificate). See `.env.example`. All values are public identifiers.

## Testing

- 12 new unit tests: session store semantics (retry swaps, scoped clears, snapshot stability) and banner state machine (delay, eligibility, confirm/fail, per-session dismissal)
- Full suite: 1348 tests pass; typecheck, lint, and `next build` clean
- Not covered: end-to-end FCM delivery — needs real Firebase credentials, recommend a preview-deploy smoke test before merging

---
## Summary by cubic
Adds push notifications for slow chat responses. After 5 seconds, users can opt in to be notified when the response is ready; handles chat switches, retries, and account changes without leaks.

- New Features
  - Offer appears after 5s during an in-progress, recovery-enabled stream; only for signed-in, non-temporary chats with push support and permission not denied.
  - Clicking Notify enables push, registers the browser with FCM via a `/push/`-scoped service worker, and creates a one-time stream watch; confirmation/failure auto-hide.
  - Tracks the active recovery session per chat and resets on chat switch or retry; guards account-switch races by scoping in-flight enables and, on sign-out, unwatching live sessions, unregistering the device, and clearing session state.
  - Adds a service worker at `public/firebase-messaging-sw.js` that suppresses notifications if the chat is visible and focuses/opens `/chat/{chatId}` on click; validates the identifier and supports local deep links (`local/{id}`).
  - Push stays off until configured; CSP updated for FCM; adds `firebase`.
  - Tests cover the session store and the banner state machine, including retry and account-switch edge cases.

- Migration
  - Set these env vars (public Firebase web config): `NEXT_PUBLIC_FIREBASE_API_KEY`, `NEXT_PUBLIC_FIREBASE_AUTH_DOMAIN`, `NEXT_PUBLIC_FIREBASE_PROJECT_ID`, `NEXT_PUBLIC_FIREBASE_MESSAGING_SENDER_ID`, `NEXT_PUBLIC_FIREBASE_APP_ID`, `NEXT_PUBLIC_FIREBASE_VAPID_KEY`.
  - Ensure controlplane endpoints support device registration and stream watches.
  - Verify CSP allows `fcmregistrations.googleapis.com` and `firebaseinstallations.googleapis.com`.
  - Do a preview deploy with real Firebase credentials to smoke test end-to-end delivery.

<sup>Written for commit f4de2107808298767c3e3f046f3ae3ffa11bc028. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/tinfoilsh/tinfoil-webapp/pull/461?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: src/services/notifications/active-stream-sessions.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
+/**
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply src/services/notifications/active-stream-sessions.ts:1]: Fixed in ddf5f117: added clearAllActiveStreamSessions() and wired it into clearAllUserData alongside resetChatRecoveryState().

File: src/services/notifications/push-notifications.ts [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+      const fcmToken = await getToken(messaging, {
+        vapidKey: FIREBASE_VAPID_KEY,
+        serviceWorkerRegistration: registration,
+      })
+      if (!fcmToken) return false
+
+      // Re-registering after a token rotation: the old token is pruned
+      // server-side when FCM reports it unregistered, so only the fresh
+      // token needs to be stored.
+      if (cachedFcmToken !== fcmToken) {
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply src/services/notifications/push-notifications.ts:98]: Fixed in ddf5f117: added resetPushNotifications() (clears cachedFcmToken and enablePromise) and wired it into clearAllUserData. firebaseApp is intentionally kept — the Firebase app instance is account-agnostic and re-initializing it per sign-in buys nothing.

File: src/components/chat/chat-interface.tsx [hunk 3]
```diff
@@ -807,6 +809,20 @@ export function ChatInterface({
...
+    bannerState: notifyBannerState,
+    requestNotification,
+    dismissBanner: dismissNotifyBanner,
+  } = useStreamNotifyBanner({
+    chatId: currentChatId ?? '',
+    // Local-only chats live under /chat/local/, so the push deep link needs
+    // the prefix to land on the right route.
+    watchChatId: currentChat?.isLocalOnly
+      ? `local/${currentChatId ?? ''}`
+      : (currentChatId ?? ''),
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply src/components/chat/chat-interface.tsx]: Rejecting: /chat/{id} is a working route for project chats — the chat loads and the app canonicalizes the URL to /project/{projectId}/chat/{id} itself via updateUrlForChat (chat-interface.tsx:1002-1008, which reads chat.projectId). Embedding the project route in the push payload would leak an extra identifier to FCM for no functional gain, which cuts against the privacy goal of this feature. Documented on getPushWatchChatId in ddf5f117.

File: src/components/chat/hooks/use-stream-notify-banner.ts [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+  const [bannerState, setBannerState] = useState<NotifyBannerState>('hidden')
+  // Sessions the user dismissed or already set a watch on; keyed by session
+  // so a retried stream (fresh session) can offer again.
+  const handledSessionsRef = useRef<Set<string>>(new Set())
+
+  const eligible =
+    Boolean(sessionId) &&
+    responsePending &&
+    pushNotificationsAvailable() &&
+    !pushPermissionDenied() &&
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply src/components/chat/hooks/use-stream-notify-banner.ts:78]: Fixed in ddf5f117: the eligibility effect now lets 'failed' (like 'confirmed') persist through its own auto-hide timeout, so denying the permission prompt shows the failure notice instead of vanishing. Covered by a new test that flips pushPermissionDenied during enable.

File: src/components/chat/hooks/use-stream-notify-banner.ts [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    !handledSessionsRef.current.has(sessionId as string)
+
+  useEffect(() => {
+    if (!eligible) {
+      setBannerState((current) => {
+        // Let the confirmation linger through its own timeout even if the
+        // stream finishes quickly; everything else hides immediately.
+        if (current === 'confirmed') return current
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply src/components/chat/hooks/use-stream-notify-banner.ts]: Fixed in ddf5f117: the banner state machine now resets to hidden whenever chatId changes, so a confirmation can't follow the user into another chat. Covered by a new chat-switch test.

File: src/components/chat/hooks/use-chat-messaging.ts [hunk 6]
```diff
@@ -902,6 +906,9 @@ export function useChatMessaging({
                       turnId,
                       sessionId,
                     )
+                    // Expose the live session so the notify banner can watch
+                    // this stream for a push notification.
+                    setActiveStreamSession(streamChatIdRef.current, sessionId)
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply src/components/chat/hooks/use-chat-messaging.ts:917]: Fixed in ddf5f117: onAttemptAbandoned now clears the published session (scoped to that session ID) before delegating to abandonChatRecoveryAttempt, so the banner can't register a watch on a dead session during retry backoff.

File: src/components/chat/hooks/use-chat-messaging.ts [hunk 7]
```diff
@@ -1220,6 +1227,7 @@ export function useChatMessaging({
           isThinking: false,
         })
         clearController(streamChatIdRef.current)
+        clearActiveStreamSession(streamChatIdRef.current)
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply src/components/chat/hooks/use-chat-messaging.ts]: Fixed in ddf5f117: each stream now remembers the session it published (publishedSessionId) and the finally-block clear is scoped to it, so a finishing stream can no longer clobber a newer stream's session for the same chat.

File: src/components/chat/hooks/use-stream-notify-banner.ts [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+    const timer = window.setTimeout(() => {
+      setBannerState('hidden')
+    }, CONSTANTS.NOTIFY_BANNER_CONFIRMATION_MS)
+    return () => window.clearTimeout(timer)
+  }, [bannerState])
+
+  const requestNotification = useCallback(() => {
+    if (!sessionId) return
+    setBannerState('enabling')
+    void (async () => {
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply src/components/chat/hooks/use-stream-notify-banner.ts:131]: Fixed in ddf5f117: requestNotification now re-reads the live session from the store after the async enable step — it watches the fresh session if the stream retried, hides quietly if the stream finished, and bails if the user switched chats. Covered by new tests for the retry-mid-enable and finished-mid-enable cases.

File: public/firebase-messaging-sw.js [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    })(),
+  )
+})
+
+self.addEventListener('notificationclick', (event) => {
+  event.notification.close()
+  const chatId = event.notification.data?.chatId
+  if (!chatId) return
+
+  const chatPath = `/chat/${chatId}`
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply public/firebase-messaging-sw.js:82]: Fixed in ddf5f117: added sanitizedChatId() enforcing ^(local/)?[0-9A-Za-z_-]{1,64}$ before any path construction, applied in both the push handler and notificationclick (data can survive SW updates, so click re-validates too).

File: src/components/chat/hooks/use-stream-notify-banner.ts [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+    if (!eligible) {
+      setBannerState((current) => {
+        // Let the confirmation linger through its own timeout even if the
+        // stream finishes quickly; everything else hides immediately.
+        if (current === 'confirmed') return current
+        return 'hidden'
+      })
+      return
+    }
+    const timer = window.setTimeout(() => {
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply src/components/chat/hooks/use-stream-notify-banner.ts:113]: Fixed in ddf5f117, via the simpler of the suggested fixes: bannerState now resets to 'hidden' when the session is replaced (and on chat switch), so each session starts a clean state machine and the offer timer's functional update can't be swallowed by a lingering confirmation.

File: tests/hooks/use-stream-notify-banner.test.tsx [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+
+  it('never offers without an active recovery session', () => {
+    const { result } = renderBanner()
+    act(() => {
+      vi.advanceTimersByTime(CONSTANTS.NOTIFY_BANNER_DELAY_MS * 2)
+    })
+    expect(result.current.bannerState).toBe('hidden')
+  })
+
+  it('never offers when push is unavailable in this browser', () => {
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply tests/hooks/use-stream-notify-banner.test.tsx:80]: Rejecting: the repeated two lines are plain arrange steps, and several tests intentionally vary them (advancing DELAY_MS-1 then 1 to probe the boundary, clearing mid-delay, different session IDs). Extracting a helper would hide the timing being tested behind an abstraction without reducing drift risk — the constant is already referenced from CONSTANTS in one place.

File: src/components/chat/chat-interface.tsx [hunk 12]
```diff
@@ -807,6 +809,20 @@ export function ChatInterface({
     currentChat?.webSearchEnabled,
   )
 
+  const {
+    bannerState: notifyBannerState,
+    requestNotification,
+    dismissBanner: dismissNotifyBanner,
+  } = useStreamNotifyBanner({
+    chatId: currentChatId ?? '',
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply src/components/chat/chat-interface.tsx:821]: Fixed in ddf5f117 (differently than suggested): the mapping moved into getPushWatchChatId() in utils/navigation.ts next to getChatPath, so the local/ route knowledge lives with the rest of the routing helpers instead of as a one-off constant at the call site.

File: src/components/chat/hooks/use-chat-messaging.ts [hunk 13]
```diff
@@ -1220,6 +1239,9 @@ export function useChatMessaging({
           isThinking: false,
         })
         clearController(streamChatIdRef.current)
+        // Scoped to our own session: a newer stream for this chat may have
+        // already published a replacement, which must survive our cleanup.
+        clearActiveStreamSession(streamChatIdRef.current, publishedSessionId)
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply src/components/chat/hooks/use-chat-messaging.ts]: Fixed in f4de2107 as suggested: the finally-block clear is now guarded on publishedSessionId !== undefined, so a stream that failed before onAttemptStarted clears nothing.

File: src/utils/signout-cleanup.ts [hunk 14]
```diff
@@ -72,6 +74,11 @@ async function clearAllUserData(options: ClearUserDataOptions): Promise<void> {
   resetTinfoilClient()
   resetChatRecoveryState()
 
+  // Drop stream-notification state: tracked live sessions and the
+  // per-account FCM device registration cache.
+  clearAllActiveStreamSessions()
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply src/utils/signout-cleanup.ts:90]: Fixed in f4de2107, with one correction to the premise: payloads contain no chat titles — only an opaque chatId and static, content-free strings, by design. But the stale-binding point stands. Signout now (1) fire-and-forget deletes every active stream watch while the auth token is still valid, and (2) resetPushNotifications() unregisters the cached device token server-side, so watches created by the signing-out account have no registered device to fan out to. SW unregistration was deliberately skipped: with the device row gone the controlplane sends nothing to this browser, and tearing down the push subscription would break other signed-in profiles' expectations if the same browser re-enables — the server-side revocation is the authoritative cut.

File: src/services/notifications/push-notifications.ts [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+}
+
+/**
+ * Forgets the in-memory registration state, e.g. on sign-out or account
+ * switch. The FCM token itself is device-scoped, but the controlplane
+ * registration is per-user, so the next account must re-register instead of
+ * hitting the previous account's already-registered cache.
+ */
+export function resetPushNotifications(): void {
+  cachedFcmToken = null
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply src/services/notifications/push-notifications.ts:163]: Fixed in f4de2107 with a generation counter rather than promise identity (promise-identity in finally trips TS use-before-assignment on the self-referencing initializer): resetPushNotifications() bumps enableGeneration; an in-flight enable re-checks its generation before writing cachedFcmToken, before reporting success, and before releasing the single-flight guard, so a stale attempt can neither clobber the next account's guard nor restore the old cache.

File: src/components/chat/hooks/use-stream-notify-banner.ts [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+
+  // Each chat gets its own banner lifecycle: switching chats resets the
+  // state machine so a confirmation can't linger into another chat.
+  useEffect(() => {
+    setBannerState('hidden')
+  }, [chatId])
+
+  // A replacement session (stream retry) also restarts the state machine,
+  // so a confirmation for the dead session can't swallow the fresh offer.
+  const previousSessionRef = useRef(sessionId)
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply src/components/chat/hooks/use-stream-notify-banner.ts]: Fixed in f4de2107: the reset effect now tracks the last non-null session (skipping the null gap instead of overwriting the ref with it), so the old -> null -> new lifecycle triggers the reset and a lingering confirmed/failed notice can't suppress the fresh session's offer. Covered by a new test that clears the store before republishing.

File: src/components/chat/hooks/use-stream-notify-banner.ts [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+      if (!enabled) {
+        setBannerState('failed')
+        return
+      }
+      if (!liveSessionId) {
+        // The stream already finished; its response is on screen.
+        setBannerState('hidden')
+        return
+      }
+      const watching = await watchStreamForPush(liveSessionId, watchChatId)
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply src/components/chat/hooks/use-stream-notify-banner.ts:149]: Fixed in f4de2107: after the watch await, the callback re-reads the live session and only confirms/fails when it still equals the watched liveSessionId; otherwise it hides quietly and lets the fresh session run its own offer cycle (the dead session is still marked handled when its watch was accepted, since acceptance means a push for it may fire). Covered by a new test that swaps the session mid-watch.

File: public/firebase-messaging-sw.js [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+
+/**
+ * Validates the chat identifier used to build the deep-link path. Chat IDs
+ * are `<digits>_<uuid>` (optionally prefixed with `local/` for local-only
+ * chats); anything else - path traversal, query strings, absolute URLs - is
+ * rejected so a forged payload cannot steer navigation.
+ */
+function sanitizedChatId(raw) {
+  if (typeof raw !== 'string') return null
+  const match = /^(local\/)?[0-9A-Za-z_-]{1,64}$/.exec(raw)
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply public/firebase-messaging-sw.js:25]: Rejecting the stricter regex, fixing the docs instead (f4de2107). The security boundary here is the URL shape, not the ID format: the current charset admits exactly one path segment under /chat/, which already forecloses traversal, query/fragment injection, and off-route navigation — a forged-but-well-formed chatId lands on the harmless not-found view whether or not it matches <digits>_<uuid>. Meanwhile chat IDs are not uniformly that shape (imports via generateChatId, legacy data), so pinning the regex to today's generator risks breaking real deep links for zero added containment. The misleading comment was the actual bug; it now documents the boundary accurately.
